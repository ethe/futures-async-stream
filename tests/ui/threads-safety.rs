#![feature(generators)]

use futures_async_stream::stream;

fn assert_send<T: Send>(_: T) {}
fn assert_sync<T: Send>(_: T) {}

#[stream(item = i32)]
pub async fn unboxed() {
    yield 0;
}

#[stream(boxed, item = i32)]
pub async fn boxed() {
    yield 0;
}

#[stream(boxed_local, item = i32)]
pub async fn boxed_local() {
    yield 0;
}

fn main() {
    assert_send(unboxed());
    assert_sync(unboxed());
    assert_send(boxed());
    assert_send(boxed_local()); //~ ERROR `dyn futures_core::stream::Stream<Item = i32>` cannot be sent between threads safely
}
