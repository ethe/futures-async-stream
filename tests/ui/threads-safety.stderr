error[E0277]: `dyn Stream<Item = i32>` cannot be sent between threads safely
  --> tests/ui/threads-safety.rs:27:17
   |
27 |     assert_send(boxed_local()); //~ ERROR `dyn futures_core::stream::Stream<Item = i32>` cannot be sent between threads safely
   |     ----------- ^^^^^^^^^^^^^ `dyn Stream<Item = i32>` cannot be sent between threads safely
   |     |
   |     required by a bound introduced by this call
   |
   = help: the trait `Send` is not implemented for `dyn Stream<Item = i32>`
   = note: required for `Unique<dyn Stream<Item = i32>>` to implement `Send`
note: required because it appears within the type `Box<dyn Stream<Item = i32>>`
  --> $RUST/alloc/src/boxed.rs
   |
   | pub struct Box<
   |            ^^^
note: required because it appears within the type `Pin<Box<dyn Stream<Item = i32>>>`
  --> $RUST/core/src/pin.rs
   |
   | pub struct Pin<P> {
   |            ^^^
note: required by a bound in `assert_send`
  --> tests/ui/threads-safety.rs:5:19
   |
5  | fn assert_send<T: Send>(_: T) {}
   |                   ^^^^ required by this bound in `assert_send`
